README.md
# iec

This package provides a suite of functions for calculating the Index of Ecological Condition (IEC), a biotic indicator of ecological health first described by Howe et al. (2007a,b) and modified by us for Giese et al. (2015). Calculation of an IEC involves two steps 1) modeling responses of species to a reference gradient or quantitative environmental stressor (typically completed by prior research) and 2) calculating IEC values for new sites based occurrences (e.g., presence/absence, abundance, frequency) of multiple species or taxonomic groups at the site. The method applies an iterative maximum likelihood approach for calculating both species response functions and IEC values. Functions for calculating the biotic responses to environmental stressors (BR models) are useful as stand-alone applications of environmental gradient analysis.


## Installation

The recommended way to install this package is using `install_github` from the [devtools](https://github.com/hadley/devtools) package.  Your computer will need to be able to compile from source.  Install [Rtools](http://cran.r-project.org/bin/windows/Rtools/) on Windows, or Xcode on Mac.  If you're on Linux, you're probably ready to go.

After installing Rtools or Xcode, install `devtools` and test your installation with the following:

```R
install.packages("devtools")
devtools::has_devel()
```

If `has_devel()` returns true, your system is ready to install from GitHub.

```R
devtools::install_github("ngwalton/iec")
```

## Using package iec

The main functions in package `iec` are `est_brc` and `est_iec`.  The first is used to estimate Biotic Response Curves (BRC) for the taxa of interest, and the second is used to estimate Index of Ecological Condition (IEC) scores for sites using the BRCs.

Use `help(package="iec")` to view the help index of all functions available in package `iec`.

Two vignettes will be included with package `iec` (but aren't completed).  These will be accessible from the help index or with one of the following:

```R
vignette("minimal_iec", package = "iec")
vignette("extended_iec", package = "iec")
```

## References
Gnass Giese, E.E., R.W. Howe, A.T. Wolf, N.A. Miller, and N.G. Walton. 2015. Sensitivity of breeding birds to the "human footprint" in western Great Lakes forest landscapes. Ecosphere 6(6):90. http://dx.doi.org/10.1890/ES14-00414.1

 
Howe, R.W., R. R. Regal, J.M. Hanowski, G.J. Niemi, N.P. Danz, and C.R. Smith.  2007a.  An index of ecological condition based on bird assemblages in Great Lakes coastal wetlands.  Journal of Great Lakes Research 33 (Special Issue 3): 93-105. 
 
Howe, R.W., R. R. Regal, G.J. Niemi, N.P. Danz, J.M. Hanowski. 2007b.  A probability-based indicator of ecological condition. Ecological Indicators 7:793-806.


NEWS.md
# iec v0.1.2

### Bug fixes

* `est_iec` now returns NA for sites where no taxa were observed when `keep_zeros = FALSE`

* When `method = "pa"`, `est_iec` now caps `pc` (expected response value) at 1

* Fixed error in `get_strat` (in `est_iec`) when `n_rep` was less than 10

* Set x and y limits to `c(0, 10)` in `plot_iec_cor`

# iec v0.1.1

* Package `iec` now passes `R CMD check`.

* Updated Gnass Giese, et. al. (2015) Ecosphere reference

* Added new function `bin` for binning of `ref_grad` (can specify number of bins or number of approx. number of records per bin)

* Relaxed error checking in `est_brc` to allow for `ref_grad` min > 0 and max < 10 to allow for binning

* `est_brc` now fails when NAs are detected in input data

* Added option to `est_iec` to score sites based on only observed species at that specific site

* Added arguments to `plot_brc` and `brc_pdf` to allow for sub titles to main and x-axis labels

* Added the following to output from `plot_brc`: Mean square error (RMSE); Pearson correlation between reference gradient and species response; Max `ylim` is plotted at no less than 1

* Removed non-linear R^2^ and Sensitivity from output of `plot_brc`

* Fixed `plot_iec_cor` to work with vector `env_grad` (was n x 1 data frame)

* Changed reference data `fish_grad` from single column data frame to vector

* Fixed typos in documentation

# iec v0.1

* First release


tests/testthat.R
library(testthat)
library(iec)

test_check("iec")


tests/testthat/test_brc_iec.R
# Tests results from est_brc and est_iec.

# Author: Nicholas G. Walton
# Created: 22 Oct 2014
# Last updated: 25 Sept 2015


context("BRC and IEC")

# file <- system.file("R", "sysdata.rda", package = "iec")
# load(file)

data(list = c("fish_grad","fish_sp"))

br <- function() {
  grad10 <- scale10(fish_grad, TRUE)
  brc <- iec::est_brc(fish_sp, grad10)
  # need to update in light of adding Direction and Magnitude to est_brc output
  brc[, -c(1, 7, length(brc))] <- round(brc[, -c(1, 7, length(brc))], digits = 4)
  brc <- brc[, -c(7, 8)]  # temporarily ignore Direction and Magnitude
  brc
}

brc <- br()

iec <- function(brc, method){
  score <- est_iec(fish_sp, brc, method, n_reps = 10)
  score[, -1] <- round(score[, -1], digits = 4)
  score
}


test_that("est_brc returns the same thing each time", {
  expect_equal_to_reference(brc, "./refs/brc_res.rds")
})

test_that("est_iec 'pa' returns the same thing each time", {
  expect_equal_to_reference(iec(brc, "pa"), "./refs/iec_pa_res.rds")
})

test_that("est_iec 'quant' returns the same thing each time", {
  expect_equal_to_reference(iec(brc, "q"), "./refs/iec_quant_res.rds")
})


R/data_doc.R
#' Fish observations
#'
#' Small dataset of fish observations.
#'
#' @format \code{fish_sp} is a data frame containing records of 4 fish species
#'   (columns) from 25 sites (rows).
#' @seealso \code{\link{fish_grad}}
'fish_sp'

#' Environmental gradient for fish sites
#'
#' Environmental gradient scores for sites in \code{\link{fish_sp}}. Larger
#' values indicate less desirable sites, so the scale will need to be inverted
#' for use with \code{\link{est_brc}}.
#'
#' @format \code{fish_grad} is a numeric vector containing
#'   environmental gradient scores from 25 sites (rows).  Site corresponds to
#'   sites (rows) in \code{\link{fish_sp}}.
#' @examples
#' data(fish_grad)
#' grad <- scale10(fish_grad, invert = TRUE)
#' # grad is now scaled to 0-10 with 10 being the least degraded.
#' summary(grad)
#' @seealso \code{\link{scale10}}, \code{\link{est_brc}}
'fish_grad'


R/brc_function.R
# This script defines the function "est_brc" which calculates Biotic Response
# functions, AKA Biotic Response Curves (BRC). The function "est_brc"
# ("iec_builder" at the time) was originally used for Erin E. G. Giese's thesis
# (2012).  Giese's thesis used the script "IEC_Builder20120328.r" which this
# current script is based on.  The purpose of this script is to calculate
# Biotic Response (BR) functions for many species or other taxa using a
# single function.
#
# Original file (used for E. Giese's thesis): "IEC_Builder20120328.r"
#   File "IEC_Builder20120328.r" was based on March Excel files but most
#   closely resembles April files.
# "est_brc" script differs from E. Giese's thesis in that observations in "sp"
# are not constrained to be probabilities.
#
# Structure of function "est_brc":
# * "Error checking" checks that "sp" is a data frame, that "ref_grad" is
#   scaled to the range 0-10, and that "sp" and "ref_grad" have the same
#   number records.
# * "Declarations" contains variables that will be used later in the
#   "est_brc".  These include the constraints placed on the optimization
#   function "nlminb".  This is where you could modify the constraints if
#   desired.
#   In general, these are the only variables you are likely to modify in this
#   function.
# * "Functions" contains function "f" which returns the lack-of-fit criterion
#   which is minimized by "nlminb".
# * "for loop over each species" contains a for loop that
#   processes each taxa in turn. "nlminb" tries to minimize the lack-of-fit
#   score returned by function "f".
#
# Authors: Nicholas G. Walton and Robert W. Howe
# Created: Mar 2012
# Last updated: 25 Sept 2015

#' Estimate Biotic Response Curves (BRC).
#'
#' \code{est_brc} generates biotic response curves (BRC) for each species or
#' taxon in \code{sp} in relation to environmental reference gradient
#' \code{ref_grad}.
#'
#' The Biotic Response Curves or functions returned by \code{est_brc} are normal
#' curves fit to the observations in \code{sp} using a lack-of-fit (LOF)
#' criteria. BRCs consist of a normal curve defined by mean (mu) and standard
#' deviation (sigma), which is multiplied (scaled) a height factor (H).  The
#' reference gradient (\code{ref_grad}) input to \code{est_brc} must be a
#' numeric vector scaled to 0-10 where 10 represents the least impacted site.
#' Use \code{\link{scale10}} to scale the reference gradient if needed. Note
#' that \code{ref_grad} must have the same order by site as \code{sp}. The
#' results of \code{est_brc} are used to give sites an Index of Ecological
#' Condition score using function \code{\link{est_iec}}.
#'
#' @param sp community data frame (sites as rows, taxa as columns, observations
#'   as values).
#' @param ref_grad numeric vector of reference gradient scores (0-10), one for
#'   each site.
#' @return A data frame defining BRCs for each species or other taxa in
#'   \code{sp}.
#' @references Gnass Giese, E.E., R.W. Howe, A.T. Wolf, N.A. Miller, and N.G.
#'   Walton. 2015. Sensitivity of breeding birds to the "human footprint" in
#'   western Great Lakes forest landscapes. Ecosphere 6(6):90.
#'   http://dx.doi.org/10.1890/ES14-00414.1
#'
#'   Howe, R.W., R. R. Regal, J.M. Hanowski, G.J. Niemi, N.P. Danz, and C.R.
#'   Smith.  2007a.  An index of ecological condition based on bird assemblages
#'   in Great Lakes coastal wetlands.  Journal of Great Lakes Research 33
#'   (Special Issue 3): 93-105.
#'
#'   Howe, R.W., R. R. Regal, G.J. Niemi, N.P. Danz, J.M. Hanowski. 2007b.  A
#'   probability-based indicator of ecological condition. Ecological Indicators
#'   7:793-806.
#' @seealso \code{\link{scale10}}
est_brc <- function(sp, ref_grad) {
  # The input "sp" is a data frame containing the observations of
  # each species or other taxa.  The row names of "sp" must
  # be site names.  All columns must contain
  # the observations of each species at each site (rows).

  # Example of "sp":
  # row.names   species(1)    species(2)    species(...)   species(n)
  # "111"           0.104         0.291             ...        0.195
  # "156"           0.266         0.391             ...        0.040
  # "116"           0.023         0.663             ...        0.199
  #  ...              ...           ...             ...          ...
  # "120"           0.693         0.737             ...        0.532

  # Input "ref_grad" in a numeric vector containing the environmental reference
  # gradient for each site. "ref_grad" must be scaled from 0 - 10 with 10 being
  # the desirable condition. "sp" and "ref_grad" must have the same order by
  # site.


  ## Error checking ----

  # Check that the input "sp" is a data frame.
  if (!is.data.frame(sp)) {
    stop("sp must be a data frame.")
  }

  # Check that input "ref_grad" is a vector.
  if (!is.vector(ref_grad)) {
    stop("ref_grad must be a vector.")
  }

  # Check for NA values.
  if (any(is.na(sp))) {
    stop("sp contains NA values.")
  }

  if (any(is.na(ref_grad))) {
    stop("ref_grad contains NA values.")
  }

  # Check that "ref_grad" is scaled correctly.
  # Checks that all "ref_grad" values are within [0, 10]
  if (min(ref_grad) < 0 | max(ref_grad) > 10) {
    stop("ref_grad is not scaled from 0 to 10.")
  }

  # Check that "ref_grad" and "sp" have the same number of records.
  if (length(ref_grad) != nrow(sp)) {
    stop("sp and ref_grad do not have the same numer of records.")
  }


  ## Declarations ----

  # Constraints on optimization - nlminb()
  mu_min <- -10  # Mean lower bound
  mu_max <-  20  # Mean upper bound
  sd_min <-   2  # SD lower bound
                 #   Results appear to be better with higher SD (was 1e-10)
  sd_max <-  10  # SD upper bound
  ht_min <-   0  # Height scaler lower bound
  ht_max <- Inf  # Height scaler upper bound

  # Generate a data frame to hold the output parameters of BR function.
  # This is an empty data frame with 6 columns - data types defined.
  brc_pars <- data.frame(character(0), rep(list(numeric(0)), 5),
                         character(0), numeric(0),
                         stringsAsFactors = FALSE)
  # Add column names.
  # Consider changing to: c("taxon", "lof", "r2", "mu", "sigma", "ht")
  names(brc_pars) <-
    c("Taxon", "LOF", "R2", "Mean", "SD", "H", "direction", "range")

  # Expand rows in brc_pars to same length as ncol in sp
  brc_pars[ncol(sp), ]  <- NA


  ## Functions ----

  f <- function(x) {
    # This function returns the Lack of Fit (LOF) score which will be
    # minimized with function "nlminb".
    # x is a numeric vector containing c(Mean, SD, H).
    # Note that ref_grad and observed are called from outside the function.
    # Consider passing from "nlminb", but may slow down processing.

    # LOF equation:
    # [(observed-expected)^2]/expected

    mu_f <- x[1]  # Mean
    sd_f <- x[2]  # Standard deviation
    ht_f <- x[3]  # A scaling factor for height of normal curve

    # Calculate the expected values (Exp) based on the current curve.
    expected <- dnorm(ref_grad, mu_f, sd_f) * ht_f

    # Calculate and return the LOF statistic.
    # Ensure that expected is 0.001 or greater.
    expected[expected < 0.001] <- 0.001  # Faster than using pmax.

    # Calculate the squared deviation for each observation.
    obs_ex2 <- (observed - expected) ^ 2

    sum(obs_ex2 / expected) # Return the Lack of Fit score.
  }


  get_strat <- function(mu_min, mu_max, sd_min, sd_max) {
    # returns a 48 x 3 data frame of stratified starting values for f

    # define a list to hold output
    start_val <- list()

    # generate breaks for mean, SD, and ht
    mu_br <- seq(from = mu_min, to = mu_max, length.out = 5) # mean breaks
    sd_br <- seq(from = sd_min, to = sd_max, length.out = 4) # sd breaks
    ht_br <- seq(from = 0, to = 50, length.out = 5)          # h breaks

    vals <- function(breaks) {
      # returns random numbers stratified by the bins specified in breaks.
      num_bins <- length(breaks) - 1
      runif(num_bins, min = breaks[1:num_bins], max = breaks[-1])
    }

    get_vals <- function(n, breaks) {
      unlist(lapply(1:n, function(x) vals(breaks)))
    }

    start_val$mu <- get_vals(12, mu_br)
    start_val$sd <- get_vals(16, sd_br)
    start_val$ht <- get_vals(12, ht_br)

    # order ht
    start_val$ht <- start_val$ht[order(rep(1:(length(ht_br) - 1), 12))]

    # return a data frame of stratified starting values for nlminb
    data.frame(start_val)
  }


  ## for loop over each species ----

  # This for loop iteratively processes each species in the data frame
  # "sp".
  for (species in 1:ncol(sp)) {

    # A list to contain the best solution found by "nlminb()".
    best <- NULL

    # Extract the observed probabilities for the current species.
    observed <- sp[, species]

    # generate random starting values
    br_start <- get_strat(mu_min, mu_max, sd_min, sd_max)

    # function to add R^2
    add_r2 <- function() {
      expect <- dnorm(ref_grad, result$par[1], result$par[2]) * result$par[3]
      iec::nlr2(observed, expect)
    }

    for (i in 1:nrow(br_start)) {

      # Run "nlminb"
      result <- nlminb(start = br_start[i, ], objective = f,
                       lower = c(mu_min, sd_min, ht_min),
                       upper = c(mu_max, sd_max, ht_max))

      # On the 1st iteration set "best" to "result".
      # After the 1st iteration, check if newest "result" is better than
      # the current best result ("best").
      if (i == 1) {
        best <- result

        # Add nonlinear R^2.
        best$R2 <- add_r2()
      } else if (result$objective < best$objective) {
        # If the new "result" is better than the current best, update
        # "best" to "result".
        best <- result
        best$R2 <- add_r2()
      }
    }

    # Add direction and range
    x <- seq(0, 10, by = 0.1)

    preds <- dnorm(x, best$par[1], best$par[2]) * best$par[3]

    # note that this will need changing if x is modified
    # also note that, as x is currently defined, 81 = 8, and 21 = 2
    # e.g., to set the boundaries to 9 and 1, use 91 and 11
    if(which.max(preds) > 81) {
      resp <- "positive"

    } else if(which.max(preds) < 21) {
      resp <- "negative"

    } else {
      resp <- "intermediate"
    }

    best$brc_dir <- resp  # response direction
    best$brc_range <- max(preds) - min(preds)

    # After fitting a best solution, add the solution to the data frame
    # "brc_pars".
    brc_pars[species, ] <- list(names(sp)[species],
                                best$objective,
                                best$R2,
                                best$par[1],
                                best$par[2],
                                best$par[3],
                                best$brc_dir,
                                best$brc_range)
  }


  ## Add sensitivity ----

  sens <- iec::get_sens(sp, ref_grad)

  brc_pars <- cbind(brc_pars, sens)


  # return brc_pars
  brc_pars
}


R/iec_function.R
# This script defines the function "est_iec" which calculates Index of
# Ecological Condition (IEC) scores. The function "est_iec"
# ("iec_site_score" at the time) was originally used for Erin E. G. Giese's
# thesis (2012).  Giese's thesis used the script
# "IEC_Application_NoForLoopExp20120328.r" which this current function is based
# on.  The purpose of this script is to estimate IEC scores based on Biotic
# Response (BR) functions of species or other taxa for many sites at once.
# BR functions can be calculated using function "est_brc".
#
# File used for E. Giese's thesis: "IEC_Application_NoForLoopExp20120328.r"
# File "IEC_Builder20120328.r" was based on March Excel files but most closely
# resembles April files (right?  based on BRC script comments).
#
# Input:
# "method" is used to set the criteria to use for estimating IEC.  "method"
# can be set to "pa" (default), or "q".  Method "pa" is used
# where present/absence data are available, and method "q" is used if
# probability data are available.
# Note that method "pa" corresponds to the Excel tab "CalcPA" (formerly
# "CalcLik"), and method "q" corresponds to Excel tab "CalcQ" (formerly
# "CalcLSq").
#
# Note: Future versions of "est_iec" may be written so that only species in
# "brc" are used for scoring sites.
#
# Structure of function "est_iec":
# * "Error checking" ensures that input variables "sp" and "brc"
#   are both data frames, and that they contain the same species.  If any of
#   these criteria are not met, an error is raised and the function exits with
#   a message describing the issue.
# * "Declarations" contains variables that will be used later in the
#   function.  In general, these are the only variables you are likely to
#   modify in this script.
# * "Main Function" contains function "f" which returns the likelihood
#   function which is minimized by "nlminb".  Note that Dr. Howe's MS Excel
#   based versions of this function maximize this function. Because "nlminb"
#   can only minimize functions, the function here has the sign reversed
#   during analysis, but the sign is corrected before adding it to the output
#   so the numbers generated are comparable to those generated by Solver in
#   Excel.
# * "for loop over each site" contains a for loop that processes
#   each taxa in turn.  Each taxa has function "nlminb" called on it the
#   number of times specified by input variable "n_reps".  "nlminb" tries to
#   minimize the likelihood function returned by function "f".
#
# Authors: Nicholas G. Walton and Robert W. Howe
#
# Created: Mar 2013(?)
# Last modified: 25 Sept 2015

#' Estimate IEC site scores.
#'
#' \code{est_iec} estimates the Index of Ecological Condition (IEC) for each
#' site in \code{sp} based on the biotic response curves (BRC) in \code{brc}.
#'
#' An Index of Ecological Condition (IEC) score is estimated for each site (row)
#' in data frame \code{sp}. This score is based on the Biotic Response functions
#' (curves) in \code{brc} (output from \code{\link{est_brc}}). The column names
#' in \code{sp} and column 1 of \code{brc} must contain the same taxa in the
#' same order.
#'
#' The default method treats observations as presence/absence (\code{method =
#' "pa"}). When using \code{pa}, it is essential to use BR functions based on
#' probability or probability-like (e.g., scaled to 0-1) observation data
#' (\code{\link{est_brc}}). Note that input records do not need to be coded as
#' presence/absence as the method simply checks for 0 or > 0.
#'
#' When quantitative observations are available and were used to generate BR
#' functions, set \code{method} to \code{"q"}. As with \code{"pa"} it is
#' important to match the type of data used to score the sites with the type of
#' data used to estimate the BR functions. For example, if the data used to
#' construct the BR functions in \code{est_brc} were probabilities, then the
#' data used to score the sites (\code{sp}) should also be probabilities If BR
#' functions were constructed using log transformed (e.g.
#' \code{\link[base]{log1p}}) count data, then the observations used to score
#' the sites in \code{est_iec} should be transformed in the same way.
#'
#' \code{n_reps} determines the number of times the
#' optimization function (\code{\link[stats]{nlminb}}) is run with different
#' starting values. \code{n_reps} can be set to any positive integer, but the
#' authors recommend against using values of less than 10 except for testing
#' purposes.  Larger values will take longer to process.  The default value is
#' 30.  If \code{keep_zeros = TRUE} (default), all taxa in \code{brc} are used
#' for scoring the sites. But if \code{keep_zeros = FALSE}, only those taxa
#' detected at a given site are used to score that site.  With the latter
#' option, the number of taxa used to score each site may vary.
#'
#' @param brc BRC data frame generated by \code{\link{est_brc}}.
#' @param method one of \code{"pa"} (default; CalcPA) or \code{"q"} (CalcQ).
#' @param n_reps scalar integer setting the number of random starts for
#'   optimization (default is 30).
#' @param keep_zeros logical indicating if all taxa are used in scoring
#'   (\code{TRUE}; default) or only detected taxa.
#' @inheritParams est_brc
#' @return A data frame containing the IEC scores for each site in \code{sp} and
#'   the corresponding likelihood of the score.
#' @references Gnass Giese, E.E., R.W. Howe, A.T. Wolf, N.A. Miller, and N.G.
#'   Walton. 2015. Sensitivity of breeding birds to the "human footprint" in
#'   western Great Lakes forest landscapes. Ecosphere 6(6):90.
#'   http://dx.doi.org/10.1890/ES14-00414.1
#'
#'   Howe, R.W., R. R. Regal, J.M. Hanowski, G.J. Niemi, N.P. Danz, and C.R.
#'   Smith.  2007a.  An index of ecological condition based on bird assemblages
#'   in Great Lakes coastal wetlands.  Journal of Great Lakes Research 33
#'   (Special Issue 3): 93-105.
#'
#'   Howe, R.W., R. R. Regal, G.J. Niemi, N.P. Danz, J.M. Hanowski. 2007b.  A
#'   probability-based indicator of ecological condition. Ecological Indicators
#'   7:793-806.
#' @seealso \code{\link{est_brc}} to generate biotic response curves.
est_iec <- function(sp, brc, method = "pa", n_reps = 30, keep_zeros = TRUE) {

  # The input "sp" is a data frame containing the observations for each
  # species or other taxa (abundance or presence/absence) at each site.
  # The column names of "sp" must be the species or other taxa.
  # All rows must contain the observations of each species at each site
  # (one column per species).  Row names are site names.

  # Example of "sp":
  # row.names  Sp(1)   Sp(2)   Sp(?)   Sp(n)
  # Site(1)       0       0       0       0
  # Site(2)       1       0       1       0
  # Site(?)       0       0       0       0
  # Site(n)       0       3       0       0


  ## Error checking ----

  # Check that the input "sp" and "brc" are both data frames.
  if (!is.data.frame(sp) | !is.data.frame(brc)) {
    stop("Inputs 'sp' and 'brc' must be data frames.")
  }

  # Check that "sp" and "brc" contain the same number of taxa.
  if (ncol(sp) != nrow(brc)) {
    stop("Species observation and BRC data frames differ in number of taxa.")
  }

  # Check that "sp" and "brc" contain the same taxa.
  if (!identical(as.character(names(sp)), as.character(brc[, 1]))) {
    stop(paste("Data frames 'sp' and 'brc' do not contain the same taxa \n",
               "or are not ordered the same."))
  }

  # Check that "method" has been set correctly.
  if (!method %in% c("q", "pa")) {
    stop("Method must be 'pa' (default) or 'q'.")
  }


  ## Declarations ----

  # Generate a data frame to hold the output IEC site score data;
  # an empty data frame with 3 columns - data type defined.
  iec_scores <- data.frame(character(0), rep(list(numeric(0)), 2),
                              stringsAsFactors = FALSE)
  names(iec_scores) <- c("Site", "IEC", "LogLik")

  # Make an original copy of "brc" if zeros are to be dropped from each site
  if (!keep_zeros) brc_raw <- brc

  # Function to generate stratified random selection of starting IEC values.
  get_strat <- function(n) {
    # Generalized to allow for other numbers to be set for "n" ("n_reps").
    # "n" does not have to be a multiple of 10.
    # Note that it's ok to set "n_reps" as low as 10,
    # and it can be set even lower if you're just running tests.

    if (n < 10) {
      int <- 10/n
      llimit <- sapply(0:(n - 1), function(x) x * int)
      strat <- sapply(llimit, function(x) runif(1, min = x, max = x + int))
    } else {
      # Lower limit for each stratum (upper will be llimit + 1).
      llimit <- 0:9

      # This is the number of random values that will be selected from
      # each stratum.
      pick <- n %/% length(llimit)

      # Select stratified random values.
      strat <- sapply(llimit, function(x) runif(pick, min = x, max = x + 1))

      # If n has been set to a value that is not a multiple of 10,
      # the remainder will be filled using random values in [0, 10].
      # Note that only this part will be used to assign starting IEC values
      # if n is set to less then length(llimit) (AKA 10).
      if (length(strat) < n) {
        remain <- n - length(strat)
        strat[(length(strat) + 1):n] <- runif(remain, min = 0, max = 10)
      }
    }
    strat   # Return the vector of starting values for "nlminb".
  }

  # Set method/criteria for estimating IEC
  # Return function for "q" or "pa" based on "method"
  # Method must be set to "q" or "pa"
  if (method == "q") {
    criteria <- function(pc, observed) {
      # Least-squares method (AKA "CalcQ").
      # Note that currently the output from this will have the
      # opposite sign of that from the Excel spread sheet.
      # This needs to be fixed, but will also require modifications to "pa".
      numerator <- (observed - pc) ^ 2
      sum(numerator / pc)   # (Obs-Exp)^2 / Exp
    }
  } else {
    criteria <- function(pc, observed) {
      # Likelihood method (AKA "pa").
      # Calculate lack-of-fit expression ("lof") for each species.
      lof <- rep(NA, length(pc))  # set lof to length of pc.

      # Set pc > 1 to 1
      pc[pc > 1] <- 1

      # If species is present at the current site, log set to Log P(C).
      lof[observed > 0] <- log10(pc)[observed > 0]

      # If species is absent, set to Log (1-P(C)).
      # Per the Excel file, it's really Log(1.0001-P(C)) to avoid log of 0.
      lof[is.na(lof)] <- log10(1.0001 - pc)[is.na(lof)]

      # Return the negative sum of the lof.
      # I'm using the negative because the original function in Excel was
      # set to maximize, but "nlminb" can only minimize a function.
      # The result is the same.
      -sum(lof)
    }
  }


  ## Main Function ----
  f <- function(x) {
    # This function returns the function which
    # will be minimized with function "nlminb".
    # "x" is the current IEC score being tried by "nlminb".
    # Note that "observed" is called from outside the function, but there
    # doesn't seem to be a better way to do this with "nlminb".

    # Calculate P(C) for each species.
    # In the original version, P(C) was constrained to > 0 and <= 1.
    # P(C) is the probability or value of each species at given IEC.
    # Input values "Mean", "SD", and "H" are part of data frame "brc".
    pc <- with(brc, {dnorm(x, mean = Mean, sd = SD) * H})
    pc[pc < .001] <- .001        # Set 0 probabilities to .001.

    # criteria is set in the Declarations section of this script.
    # It is set to either "pa" (default) or "q".
    criteria(pc, observed)
  }


  ## for loop over each site ----
  # This for loop iteratively processes each site in the data frame "sp".

  for (site in 1:nrow(sp)) {
    # A list to contain the best solution found by "nlminb".
    best <- NULL

    # Extract the "observed" responses for the current site.
    # If "keep_zeros" is TRUE (default), all taxa are used for scoring.
    # Else, only those taxa that were detected are used for scoring.
    if (keep_zeros) {
      observed <- sp[site, ]
    } else {
      # "keep" contains the indices for detected species.
      keep <- which(sp[site, ] > 0)

      # If there are no species detected, set iec attributes to NA and
      # go to the next site
      if (length(keep) == 0) {
        iec_scores[site, ] <- list(row.names(sp)[site], NA_real_, NA_real_)
        next
      }

      # subset "observed" and "brc" to the detected species
      observed <- sp[site, keep]
      brc <- brc_raw[keep, ]
    }


    # Function "get_strat" is defined in the "Declarations" section.
    iec_start <- get_strat(n_reps)

    # Estimate IEC for current site using "n_reps" different starting values.
    for (i in 1:n_reps) {
      # "n_reps" is set in the call to function "siteScore" and
      # defaults to 30.

      # For help with function "nlminb" type "?nlminb" in R.
      # This is the optimization function.
      # The result is constrained to be between 0 and 10 inclusive.
      result <- nlminb(start = iec_start[i], objective = f,
                       lower = 0, upper = 10)

      # Check if the current random starting value resulted in an
      # improved estimate of IEC.
      if (i == 1) {
        # On the 1st iteration set "best" to "result".
        best <- result
      } else if (result$objective < best$objective) {
        # Otherwise check if newest "result" is better than the current
        # best result ("best").
        # If the new "result" is better than the current best,
        # update "best" to "result".
        best <- result
      }
    }

    # After fitting a best solution, add the solution to the
    # data frame "iec_scores".
    # best$objective is multiplied by -1 to put it on the same scale as
    # the original Excel based versions.
    iec_scores[site, ] <- list(row.names(sp)[site], best$par, -best$objective)
  }

  # Return "iec_scores"
  iec_scores
}
